source/module_hsolver/diago_david.cpp

```diff
@@ -781,7 +784,7 @@ void DiagoDavid<T, Device>::refresh(const int& dim,
     basis.zero_out();
     for (int m = 0; m < nband; m++)
     {
-        syncmem_complex_op()(this->ctx, this->ctx, &basis(m, 0), &psi(m, 0), this->dim);
+        syncmem_complex_op()(this->ctx, this->ctx, &basis(m, 0), psi.get_k_first() ? &psi(m, 0) : &psi(m, 0, 0), this->dim);
```

Comment:
Can you refactor Psi::fix_k() to skip k index without k_first?

Reply:
No need to skip. k_first=0 means that nks is a dimension of **basis**  to be contracted in "HPsi". In LR-TDDFT the formula writes: 
$$\sum_{jb\mathbf{k}'}A^I_{ai\mathbf{k}, jb\mathbf{k}'}X^I_{jb\mathbf{k}'}$$
In the code: 
-  H means A
- Psi means X
- "band" means I (the number of excited states to be solved)
- k: k-points, same meaning as the ground state
- "basis": nocc(i,j in formulas)*nvirt(a,b in formulas), corresponding to "bands" of the ground state

Reply:
> No need to skip. k_first=0 means that nks is a dimension of **basis** to be contracted in "HPsi". In LR-TDDFT the formula writes: ∑jbk′Aaik,jbk′IXjbk′I In the code:
> 
> * H means A
> * Psi means X
> * "band" means the superscript I (the number of excited states to be solved)
> * k: k-points, same meaning as the ground state
> * "basis": number of occupied ks-orbitals (subscripts i,j) * number of unoccupied ks-orbitals (subscripts a,b), corresponding to "bands" of the ground state

Can you add these comments into this code file? I think other developers will also confused with k_first related code.

Reply:
In principle, DiagoDavid only do diagonalization for target Psi which contains matrix with size of nbands * nbasis.
For LR-TDDFT, the better practice would be use a new Psi with 1 * nbands * (nks*nbasis) as a wrapper of current Psi with nbands * nks * nbands. 
The DiagoDavid class should not treat different DFT algorithm, it is only a basic math tool.

Reply:
I've added them at the place setting "dim" in DiagoDavid.

Reply:
> In principle, DiagoDavid only do diagonalization for target Psi which contains matrix with size of nbands * nbasis. For LR-TDDFT, the better practice would be use a new Psi with 1 * nbands * (nks*nbasis) as a wrapper of current Psi with nbands * nks * nbands. The DiagoDavid class should not treat different DFT algorithm, it is only a basic math tool.

I think `DiagoDavid` is not a basic math tool unless it does not depend on the class `Hamilt` and `Psi`. Now it depends on `Psi`, then using any logic implemented in Psi is reasonable. 
If we need to refactor `Diago` class with no `Hamilt` and `Psi` in the interface (such as HPC friendly pointer-only interface) ,  the k-first logic would be natually move out of `DiagoDavid`.